### ISSUE TITLE
Automatic plugin testing

### ISSUE THREAD

**Issue by madmaxoft** [cuberite/cuberite#3321]
I was wondering about the possibility of automatically testing plugins in a CI environment, at least in some way.

It should be possible to add an API dump capability to ToLua++ and / ord APIDump plugin in such a form that the results can be used to reconstruct dummy API functions implemented purely in Lua.

Then the framework could perform some automated tests of the plugin functionality:
- Load the plugin files (thus checking for syntax errors)
- Call each command handler declared in Info.lua (where to get params?)
- Call each hook registered in the `Initialize()` function / in main chunk
- Call each webadmin page handler registered in the `Initialize()` function / in main chunk
- possibly others

It might even be possible to check against storing an object for later use in a callback (typically, a `cPlayer` instance stored from a hook used in a callback later on) - the objects given to hooks can be "reset" to report usage as errors before the later callback is called.

This is most likely best implemented as a project in a separate repo, probably implemented fully in Lua; the API data could be stored online on `apidocs.cuberite.org`.

**Comment by NiLSPACE** [comment]
Is it worth it though? The APIDump plugin usually only gets a few additions or removals in case an API is added. Perhaps we can setup something plugin specific for the bigger plugins like Core, but that should be tested in their own repo's.

We could also try to get some unit-tests going - make the plugins test themselves.

**Comment by madmaxoft** [comment]
This is exactly about the generic self-tests for any plugin. If we emulate the Cuberite API in a smart way, it should be possible to catch some really interesting errors, such as that storing a volatile object  for later use.

**Comment by Cl1608Ho** [comment]
Just asking, why do you need to emulate the api? Without real players cuberite should be pretty efficient so you only have to simulate a fake player or two for functions which apply on players. 

Additionally, some status report, ie messages like 'ended up teleportiert player x to coordinates x,y,z other players locations:....'

Or if cBlockArea was used, render what was left (=what is now in the world) using your schematics renderer @madmaxoft (the one you used for the gallery)

Basically it's a before-to-after report between command usages. For which parameters and such I would recommend a testing parameter. Otherwise only style checks 

And do not forget events, at least two times for each testing scenario (like change block or whatever two times for one block but not for every block :P if someone uses the first call differently)

**Comment by NiLSPACE** [comment]
We can't simulate things like cPlayer objects with that because they are too tightly coupled with cClientHandle which requires a cProtocol class which requires a network connection.

**Comment by madmaxoft** [comment]
Not only that, but there are things that Cuberite itself cannot detect - such as that constantly feared storing a volatile object from a callback for later use.

**Comment by NiLSPACE** [comment]
We can test if something is saved by using the `newproxy` function and giving it a `__gc` metamethod. After we simulate something like a hook call, or a command call we fire `collectgarbage`, and if it's collected it's fine, but if not it was probably saved somewhere.

**Comment by madmaxoft** [comment]
Even easier, after each hook call, remove all the object's functions and variables, this way if it is used later on, Lua will abort because the plugin will be trying to call or access a `nil`. If we wanted to keep going despite the error, we could add an `__index` metamethod that would create a temporary function for each access that would log the error and remove itself again. Isn't Lua just wonderfully powerful? :)

**Comment by NiLSPACE** [comment]
I thought of that as well, but then you only know if it was saved after an X amount of time/calls. If we want to quickly test it then using the `__gc` method you'd know it instantly.

**Comment by NiLSPACE** [comment]
I did a quick test to see if it worked, and it did. It noticed that the object was collected and thus it was fine. It even noticed if I created a function that was using the object later on inside the callback.

**Comment by madmaxoft** [comment]
Why would you say so? In my opinion, the simulator would work perfectly with this:
1. Create a "callback request queue", populate it with a call to `Initialize()`.
2. Take one callback request from the queue
3. Create "DummyAPI" objects for the callback parameters
4. Call the callback. Whenever the plugin asks to be called back (command reg, hook reg, schedule task etc.), store the callback request in the callback queue
5. Invalidate each "DummyAPI" object from step 3
6. Take next callback from the queue and continue with step 3

This has captured the minimum guaranteed object lifetime - the objects passed to callbacks are guaranteed to be valid only as long as the callback executes. By using a callback queue, the simulator can make sure only the objects that are supposed to be valid are actually used.

**Comment by NiLSPACE** [comment]
But still, you don't know for certain if the plugin will use the object later on somewhere else. You only know once the plugin uses the 'invalid' object somewhere else.

**Comment by madmaxoft** [comment]
Of course, we can't test all the code paths, but it will work for most instances - usually the error is that the plugin stashes the object and then use it in the callback immediately.

I'm afraid calling garbage collection would make the whole thing run too slow, and I'm not too sure if gc is allowed to leave some garbage behind for later, so it might not be reliable.

Or perhaps the two approaches can be combined?

**Comment by NiLSPACE** [comment]
Since we're only calling the garbage collector after an event/hook I doubt it would be slow enough to actually see much of a difference.

My tests were pretty reliable, though of course, the testfile was only 4 lines long. It might be different for larger plugins.

Another complication I can think of when using `__gc` is that we can't save the object ourselves. Meaning that we'd have to re-create an object for a player, or entity every time we'd call a hook. Though we could probably fix this by storing the object once the `__gc` method gets called. The proxy would be destroyed, but the object still exists.

Perhaps we could combine the approaches, but I don't think it would be pretty. Perhaps some more people can get involved in this, especially the Lua devs. @cuberite/plugins, @cuberite/core-team?

**Comment by madmaxoft** [comment]
I've got the basic framework working, it is now capable of simulating a very minimal plugin (`LOG` and `cPluginManager:BindCommand` API calls). I've written the simulator in such a way that it can provide multiple means of checking the callbacks, as well as other metrics (it could simply add a list of API functions called by the plugin, for example).

I'm developing it in a separate repo, https://github.com/madmaxoft/CuberitePluginChecker for now. If it ever reaches a mature state, we can move the repo into the Cuberite organization

**Comment by madmaxoft** [comment]
I don't like having to use undocumented API functions, but in this case there doesn't seem to be a reasonable workaround. I implemented the GC check in the checker. I'm happy to say that the Gallery plugin passed this check with all its command-handlers :)

**Comment by NiLSPACE** [comment]
That's awesome :D

**Comment by madmaxoft** [comment]
And now it can check using both methods. I have to say the GC-based method is awesome in detecting the problems, so unless it runs into performance problems (which it doesn't seem), it will be prefered over the Clear method.

Now I'd like to make sure the simulator can process the entire Core plugin. Then I'll call it complete :)

**Comment by NiLSPACE** [comment]
I had a feeling it would be great to have GC detection implemented :)

**Comment by madmaxoft** [comment]
Oh wow, the Core is actually full of errors. Arithmetic being performed on strings, strings being saved into integers, using undocumented API function overloads...

**Comment by madmaxoft** [comment]
I believe the CuberitePluginChecker project is now quite good to be used as CI for plugins.
https://github.com/cuberite/CuberitePluginChecker

Closing.
